### ISSUE TITLE
explicit_auto_deref false positive for function returning associated dyn type

### ISSUE THREAD

**Issue by yuja** [rust-lang/rust-clippy#10384]
### Summary

explicit_auto_deref suggested for `fn _(&self) -> &Self::Ty { *self }`, but the suggested code doesn't compile.

### Lint Name

explicit_auto_deref

### Reproducer

I tried this code:

```rust
trait Object {}

trait AsObject {
    type Ty: Object + ?Sized;
    fn as_object(&self) -> &Self::Ty;
}

impl<'a> AsObject for &'a dyn Object {
    type Ty = dyn Object + 'a;
    fn as_object(&self) -> &Self::Ty {
        *self
    }
}
```

I saw this happen:

```
warning: deref which would be done by auto-deref
  --> src/main.rs:11:9
   |
11 |         *self
   |         ^^^^^ help: try this: `self`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#explicit_auto_deref
   = note: `#[warn(clippy::explicit_auto_deref)]` on by default
```

The suggested code failed to compile:
```
error[E0277]: the trait bound `&'a (dyn Object + 'a): Object` is not satisfied
  --> src/main.rs:11:9
   |
11 |         self
   |         ^^^^ the trait `Object` is not implemented for `&'a (dyn Object + 'a)`
   |
   = note: required for the cast from `&'a (dyn Object + 'a)` to the object type `dyn Object`
```

If the return value was spelled without `Self::Ty`, clippy didn't emit false-positive warning.
```rust
impl<'a> AsObject for &'a dyn Object {
    type Ty = dyn Object + 'a;
    fn as_object(&self) -> &(dyn Object + 'a) {
        *self
    }
}
```

### Version

```text
rustc 1.67.1 (d5a82bbd2 2023-02-07)
binary: rustc
commit-hash: d5a82bbd26e1ad8b7401f6a718a9c57c96905483
commit-date: 2023-02-07
host: x86_64-unknown-linux-gnu
release: 1.67.1
LLVM version: 15.0.6
```

### Additional Labels

@rustbot label +I-suggestion-causes-error
